### ISSUE TITLE
GradientTape.gradient error if tf.slice is used

### ISSUE THREAD

**Issue by MPnoy** [SciSharp/TensorFlow.NET#800]
// Sample data
var X = np.array(3.3f, 4.4f, 5.5f, 6.71f, 6.93f, 4.168f, 9.779f, 6.182f, 7.59f, 2.167f,
             7.042f, 10.791f, 5.313f, 7.997f, 5.654f, 9.27f, 3.1f);
var Y = np.array(1.7f, 2.76f, 2.09f, 3.19f, 1.694f, 1.573f, 3.366f, 2.596f, 2.53f, 1.221f,
             2.827f, 3.465f, 1.65f, 2.904f, 2.42f, 2.94f, 1.3f);
var n_samples = X.shape[0];

// We can set a fixed init value in order to demo
var W = tf.Variable(-0.06f, name: "weight");
var b = tf.Variable(-0.73f, name: "bias");

// Wrap computation inside a GradientTape for automatic differentiation.
using var g = tf.GradientTape();
// Linear regression (Wx + b).
var pred = W * X + b;

// ------ Do nothing ------
var test = tf.slice(pred, new[] { 0 }, pred.shape);

// Mean square error.
var loss = tf.reduce_sum(tf.pow(test - Y, 2)) / (2 * n_samples);
// should stop recording
// Compute gradients.

// Error
var gradients = g.gradient(loss, (W, b));

**Comment by Oceania2018** [comment]
@MPnoy Could you PR the Unit Test here? https://github.com/SciSharp/TensorFlow.NET/blob/master/test/TensorFlowNET.UnitTest/ManagedAPI/GradientTest.cs

**Comment by MPnoy** [comment]
Done

**Comment by MPnoy** [comment]
Also I found that GradientTape.gradient after tf.concat always returns (null, null)

**Comment by Oceania2018** [comment]
@MPnoy I got `(None, None)` in python:
![image](https://user-images.githubusercontent.com/1705364/115795358-641cdd80-a395-11eb-83f7-405dd5e88cbf.png)

**Comment by MPnoy** [comment]
@Oceania2018 Thanks for the fix, it works.
About the concat: there is a issue in the tensorflow itself. https://github.com/tensorflow/tensorflow/issues/37726
Except for the rest, there is one working code there: https://github.com/tensorflow/tensorflow/issues/37726#issuecomment-768631785

I rewrote this into unit test and PR if necessary
https://github.com/SciSharp/TensorFlow.NET/commit/83bd764b6af982c8e3a71d1b1001b215e9f5590e

**Comment by Oceania2018** [comment]
@MPnoy Can you PR the runnable one?
